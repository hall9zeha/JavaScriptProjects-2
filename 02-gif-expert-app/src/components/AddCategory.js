import React, {useState} from 'react'
import PropTypes from 'prop-types';


export const AddCategory = ({setCategories}) => {
  //obtenemos la función enviada desde El componente GifExpertApp

  //Para que no nos de un error de undefined lo ponemos como vacío
  const [inputValue, setInputValue] = useState('');

  const handleInputChange =(e) =>{setInputValue(e.target.value);}

  const handleOnsubmit = (e)=>{
    e.preventDefault();
    if(inputValue.trim().length > 1){
    setCategories(cats=>[inputValue,...cats]);
    setInputValue('');
    }
    
  }
  return (
    // <> 
    // Si ya tenemos un contenedor de todos nuestros elementos como será el caso de nuestro formulario
    // ya no necesitaremos el fragment
    <form onSubmit={handleOnsubmit}>
     <input
        type='text'
        placeHolder='Search'
        value={inputValue}
        onChange={handleInputChange}
    />
    </form>
    //</>
    
  )
}

AddCategory.propTypes={
    setCategories:PropTypes.func.isRequired
}
