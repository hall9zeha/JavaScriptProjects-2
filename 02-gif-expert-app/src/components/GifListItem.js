import React from 'react'
import PropTypes from 'prop-types'

export const GifListItem = ({title, url}) => {
  return (
    <div  data-testid="div" className="card animate__animated animate__bounceInLeft animate__delay-700s" >
        <img src={url} alt={title}/>
        <p>{title}</p>
        </div>
  )
}

GifListItem.propTypes={
  title:PropTypes.string.isRequired,
  url:PropTypes.string.isRequired
}