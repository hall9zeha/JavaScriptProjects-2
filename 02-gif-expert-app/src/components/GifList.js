import React, {useState, useEffect} from 'react'
import { GifListItem } from './GifListItem';

export const GifList = ({category}) => {

    const [images, setImages] = useState([]);

    useEffect(()=>{
        fetchGifs();
    },[]); //Esto hará que la función sea llamada solo la primera vez que se renderice el componente

    const fetchGifs = async() =>{
        const url='https://api.giphy.com/v1/gifs/search?q=one+punch&limit=10&api_key=you-api-key';
        const resp = await fetch(url);
        const {data} = await resp.json();
        
        const gif = data.map(img=>{
            return {
                id:img.id,
                title:img.title,
                url:img.images.downsized_medium.url
            }
        });
        
        setImages(gif);
        console.log(images);
    }

  return (
   <>
    <h3>{category}</h3>
    <div className='card-grid'>
        {
            images.map( image =>
                <GifListItem 
                key = {image.id}
                {...image}/>
             
            )
        }
   
    </div>
    </>
  )
}
