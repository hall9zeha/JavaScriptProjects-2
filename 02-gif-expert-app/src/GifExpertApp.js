
import React, { memo, useState } from 'react'
import AddCategory from './components/AddCategory';
import { GifList } from './components/GifList';



const GifExpertApp = memo(() => {

  const [categories, setCategory] = useState(['One punch']);

  
  // const handleAdd = (e)=>{
  //   setCategory([...categories,e]);//mantenemos el arreglo anterior y agregamos la nueva categoría al presionar el botón
  // }

  return (
    <>

    <h2>GifExpertApp</h2>
    <AddCategory setCategories={setCategory}/>
    <hr/>
    
    <ol>
      {
        categories.map(category =>
           <GifList 
           key={category}
           category={category}/>
        )
      }
    </ol>
    </>
    
  )
})

export default GifExpertApp