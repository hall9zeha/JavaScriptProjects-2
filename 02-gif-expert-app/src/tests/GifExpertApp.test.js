import React from 'react';
import renderer from 'react-test-renderer';
import { GifList } from '../components/GifList';

import GifExpertApp from '../GifExpertApp';

describe('test in <GifExpertApp/> component', () => { 
    
    const wrapper = renderer.create(<GifExpertApp/>).toJSON()

    test('should show successful', () => { 
        expect(wrapper).toMatchSnapshot();
    });

    test('should show a category list',()=>{
        const categories = ['Boku no hero', 'Komi san'];
        const wrapper = renderer.create(<GifExpertApp defaultCategories={categories}/>);

        expect(wrapper).toMatchSnapshot();
        const gridGifs = wrapper.root.findAllByProps({className:'card-grid'});
        expect(gridGifs).toHaveLength(categories.length)
    });
})