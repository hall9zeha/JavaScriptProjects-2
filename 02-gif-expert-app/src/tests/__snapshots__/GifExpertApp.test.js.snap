// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`test in <GifExpertApp/> component should show a category list 1`] = `
Array [
  <h2>
    GifExpertApp
  </h2>,
  <form
    onSubmit={[Function]}
  >
    <input
      onChange={[Function]}
      type="text"
      value=""
    />
  </form>,
  <hr />,
  <ol>
    <h3>
      Boku no hero
    </h3>
    <p
      className="loading"
    >
      Cargando...
    </p>
    <div
      className="card-grid"
    />
    <h3>
      Komi san
    </h3>
    <p
      className="loading"
    >
      Cargando...
    </p>
    <div
      className="card-grid"
    />
  </ol>,
]
`;

exports[`test in <GifExpertApp/> component should show successful 1`] = `
Array [
  <h2>
    GifExpertApp
  </h2>,
  <form
    onSubmit={[Function]}
  >
    <input
      onChange={[Function]}
      type="text"
      value=""
    />
  </form>,
  <hr />,
  <ol />,
]
`;
